pkg/wallet/wallet_test.go

```diff
@@ -48,33 +56,188 @@ func TestBaseWallet_CreateKey(t *testing.T) {
+		require.NoError(t, err)
+		_, err = w.UnpackMessage(packMsg)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "get error")
+	})
+
+	t.Run("test failed to unmarshal encMessage", func(t *testing.T) {
+		w, err := New(&mockstorage.MockStoreProvider{Store: mockstorage.MockStore{
+			Store: make(map[string][]byte),
+		}}, nil)
 		require.NoError(t, err)
 		_, err = w.UnpackMessage(nil)
 		require.Error(t, err)
+		require.Contains(t, err.Error(), "failed to unmarshal encMessage")
+	})
+
+	t.Run("test key not found", func(t *testing.T) {
+		crypter, err := authcrypt.New(authcrypt.XC20P)
+		require.NoError(t, err)
+		w, err := New(&mockstorage.MockStoreProvider{Store: mockstorage.MockStore{
+			Store: make(map[string][]byte),
+		}}, crypter)
+		require.NoError(t, err)
+
+		pub1, priv1, err := box.GenerateKey(rand.Reader)
+		require.NoError(t, err)
+		base58FromVerKey := base58.Encode(pub1[:])
+		require.NoError(t, w.persistKey(base58FromVerKey, &crypto.KeyPair{Pub: pub1,
+			Priv: priv1}))
+
+		pub2, _, err := box.GenerateKey(rand.Reader)
+		require.NoError(t, err)
+
+		packMsg, err := w.PackMessage(&Envelope{Message: []byte("msg1"),
+			FromVerKey: base58FromVerKey,
+			ToVerKeys:  []string{base58.Encode(pub2[:])}})
+		require.NoError(t, err)
+
+		_, err = w.UnpackMessage(packMsg)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "no corresponding recipient key found in")
+	})
+
+	t.Run("test decrypt failed", func(t *testing.T) {
+		crypter, err := authcrypt.New(authcrypt.XC20P)
+		require.NoError(t, err)
+		w, err := New(&mockstorage.MockStoreProvider{Store: mockstorage.MockStore{
+			Store: make(map[string][]byte),
+		}}, &didcomm.MockAuthCrypt{DecryptValue: func(envelope []byte, recipientKeyPair crypto.KeyPair) ([]byte, error) {
+			return nil, fmt.Errorf("decrypt error")
+		}, EncryptValue: func(payload []byte, sender crypto.KeyPair, recipients []*[crypto.KeySize]byte) (bytes []byte, e error) {
+			return crypter.Encrypt(payload, sender, recipients)
+		}})
+		require.NoError(t, err)
+
+		pub1, priv1, err := box.GenerateKey(rand.Reader)
+		require.NoError(t, err)
+		base58FromVerKey := base58.Encode(pub1[:])
+		require.NoError(t, w.persistKey(base58FromVerKey, &crypto.KeyPair{Pub: pub1,
+			Priv: priv1}))
+
+		pub2, priv2, err := box.GenerateKey(rand.Reader)
+		require.NoError(t, err)
+		require.NoError(t, w.persistKey(base58.Encode(pub2[:]), &crypto.KeyPair{Pub: pub2,
+			Priv: priv2}))
+
+		packMsg, err := w.PackMessage(&Envelope{Message: []byte("msg1"),
+			FromVerKey: base58FromVerKey,
+			ToVerKeys:  []string{base58.Encode(pub2[:])}})
+		require.NoError(t, err)
+
+		_, err = w.UnpackMessage(packMsg)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "decrypt error")
+
 	})
 }
 
 func TestBaseWallet_PackMessage(t *testing.T) {
-	t.Run("test error not implemented", func(t *testing.T) {
-		w, err := New(&mockstorage.MockStoreProvider{})
+	t.Run("test success", func(t *testing.T) {
```

Comment:
Do we have an expected test test case for Pack and Unpack (e.g., from RFC) that we are comparing against?

Reply:
to test success we just pack msg and unpack it. I'm not sure how else can I test success

Reply:
We would need a test fixture from the RFC and ensure we produce the same. We may have issues when we try to interop with another agent if we only test against our own implementation.